Intrinsics analyzer and fixes

davidwrighton opened the pull request:
As a follow up to #85275, implement an analyzer and an attribute the Crossgen2 will respect which will make it difficult to encounter unexpected behavior at runtime due to quirks of how we compile System.Private.CoreLib.dll with Crossgen2.

See the change to vectors-and-intrinsics.md for details of what the new system will check.

NOTE: I believe that functionally this change is complete. I need to check crossgen2 performance, and the name of the new attribute is terrible. I'm looking for better names.

davidwrighton committed:
Implement analyzer for platform intrinsics use in System.Private.CoreLib

This analyzer detects the use of all platform intrinsics and checks to
ensure that they are all used either protected by an if statement OR
ternary operator which checks an appropriate IsSupported flag, or that
the intrinsic is used within a method where the behavior of platform
support for the intrinsic is not allowed to vary between compile time
and runtime.

The supported conditional checks are

1. Simple if statement checking IsSupported flag surrounding usage
```
if (PlatformIntrinsicType.IsSupported)
{
    PlatformIntrinsicType.IntrinsicMethod();
}
```

2. If statement check checking a platform intrinsic type which implies
that the intrinsic used is supported.

```
if (Avx2.X64.IsSupported)
{
    Avx2.IntrinsicMethod();
}
```

3. Nested if statement where there is an outer condition which is an
OR'd together series of IsSupported checks for mutually exclusive
conditions and where the inner check is an else clause where some checks
are excluded from applying.

```
if (Avx2.IsSupported || ArmBase.IsSupported)
{
    if (Avx2.IsSupported)
    {
        // Do something
    }
    else
    {
        ArmBase.IntrinsicMethod();
    }
}
```

In addtion, this change adds a new attribute, currently named
System.Runtime.BypassReadyToRunForIntrinsicsHelperUse which can be used
to identify methods which are called expecting to use some particular
set of platform intrinsics. These functions do not need to have explicit
checks, and while this is not yet implemented in the change, the
attribute will change the behavior of crossgen2 so that code will only
be generated if the platform intrinsic support is well known to be
either enabled or disabled at runtime. In addition, the method will only
be compiled or be inlineable if the platform intrinsic support is
enabled that matches at least one of the attributes applied to that
method.

davidwrighton committed:
Remove BypassReadyToRunAttribute as it should no longer be needed for intrinsics usages

davidwrighton committed:
Document new intrinsic handling behavior and enhance crossgen2 to respect the new attribute

davidwrighton committed:
Fix build break

jkoritzinsky commented on src/libraries/System.Private.CoreLib/gen/IntrinsicsInSystemPrivateCoreLibAnalyzer.cs:
```diff
@@ -0,0 +1,533 @@
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+// This isn't a shipping analyzer, so we don't need release tracking
+#pragma warning disable RS2008
```
I believe we disable Roslyn-style release tracking in this repo, so we can remove this suppression.

```suggestion
```

davidwrighton replied:
Nope. It fails the build without disabling this warning.

jkoritzinsky commented on src/libraries/System.Private.CoreLib/gen/IntrinsicsInSystemPrivateCoreLibAnalyzer.cs:
```diff
@@ -0,0 +1,533 @@
+    {
+        public const string DiagnosticId = "IntrinsicsInSystemPrivateCoreLib";
+
+        // You can change these strings in the Resources.resx file. If you do not want your analyzer to be localize-able, you can use regular strings for Title and MessageFormat.
+        // See https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Localizing%20Analyzers.md for more on localization
```
```suggestion
```

jkoritzinsky commented on src/libraries/System.Private.CoreLib/gen/IntrinsicsInSystemPrivateCoreLibAnalyzer.cs:
```diff
@@ -0,0 +1,533 @@
+                                                INamedTypeSymbol? bypassReadyToRunForIntrinsicsHelperUse)
+            {
+                NamedTypesToBeProtected = namedTypesToBeProtected;
+                BypassReadyToRunAttribute = bypassReadyToRunAttribute;
```
It looks like this PR also deletes the `BypassReadyToRunAttribute` type. Is this path that checks for it in the analyzer used?

davidwrighton replied:
It isn't, but the attribute is still there and effective. In case we need it, I don't want to rev the analyzer.

jkoritzinsky reviewed (commented):
Can we add unit tests for this just to make it easier to iterate on in the future? Since it's in the libraries test tree, it should be pretty easy to add them.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs:
```diff
@@ -409,6 +413,11 @@ public sealed override bool CanInline(MethodDesc callerMethod, MethodDesc callee
 
+            if (canInline)
+            {
+                if (!CorInfoImpl.ShouldCodeNotBeCompiledIntoFinalImage(_instructionSetSupport, calleeMethod))
```
*facepalm*

davidwrighton committed:
Fix condition noted in code review

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/BypassReadyToRunForIntrinsicsHelperUse.cs:
```diff
@@ -0,0 +1,18 @@
+    // Use this attribute to indicate that a function should only be compiled into a Ready2Run
+    // binary if the associated type will always have a well defined value for its IsSupported property
+    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor, AllowMultiple = true, Inherited = false)]
+    internal sealed class BypassReadyToRunForIntrinsicsHelperUseAttribute : Attribute
```
I like `RequiresIntrinsics` or maybe some variation there-of.

We could also maybe call it the same thing as we do in the JIT: `CompExactlyDependsOn(typeof(Avx))`?

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyValues/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -809,6 +809,9 @@ internal static int LastIndexOfAnyVectorized<TNegator>(ref byte searchSpace, int
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [System.Runtime.BypassReadyToRunForIntrinsicsHelperUse(typeof(Sse))]
+        [System.Runtime.BypassReadyToRunForIntrinsicsHelperUse(typeof(AdvSimd))]
+        [System.Runtime.BypassReadyToRunForIntrinsicsHelperUse(typeof(PackedSimd))]
```
Nit:
```suggestion
        [BypassReadyToRunForIntrinsicsHelperUse(typeof(Sse))]
        [BypassReadyToRunForIntrinsicsHelperUse(typeof(AdvSimd))]
        [BypassReadyToRunForIntrinsicsHelperUse(typeof(PackedSimd))]
```

Should it be `Sse2` instead of `Sse` here?

davidwrighton replied:
Yes. This is actually showing one of the limitations of the analyzer, which is that if you follow the pattern of having a helper which uses multiple variants which are explicitly checked, it isn't good at requiring the exact right attribute.

davidwrighton replied:
I should be able to make the analyzer detect this situation and issue a slightly different diagnostic, which could be suppressed with a pragma warning disable for legit cases.

davidwrighton commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs:
```diff
@@ -826,7 +827,6 @@ private static int ComputeFirstIndex(ref short searchSpace, ref short current, V
 
-        [BypassReadyToRun]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private static int ComputeFirstIndex(ref short searchSpace, ref short current, Vector256<byte> equals)
```
Not at the moment. I could make it so that calling an method marked with `BypassReadyToRunForIntrinsicsHelperUse` required the IsSupported checks or being called from within an appropriate `BypassReadyToRunForIntrinsicsHelperUse` method. That would make the analyzer do the "right" thing, and also make sure the helpers are called safely. Such a new feature might cause issues for the annotation on `Vector128.ShuffleUnsafe`, but possibly that's not a bad thing. That api is... odd.

davidwrighton committed:
Update analyzer to detect more patterns, and attribute much of System.Private.CoreLib

davidwrighton committed:
Add pragmas and attributes to the packed span helpers
- These use a peculiar construct but its legal as its all containing in 1 method. If it got more complex than that, the analysis would be quite a bit harder

davidwrighton committed:
Hah! It works
- Handle intrinsics calls from nested intrinsics types
- Handle IsSupported checks in the presence of an attribute which explicitly allows an IsSupported check as well as one which doesn't, but is related
- Add suppressions for Avx2 IsSupported checks in IndexOfAnyAsciiSearcher helper functions

davidwrighton committed:
Rename attribute and fix docs

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into IntrinsicsAnalyzerAndFixes

davidwrighton committed:
Add analyzer tests and fixup issues from merging with main branch

davidwrighton committed:
Disable tests on Mono due to Mono runtime bug

davidwrighton committed:
Handle incorrect namespace for the CompExactlyDependsOnAttribute and don't pass a null InstructionSetSupport into the ShouldSkipCompilation logic, just feed it around everywhere.

davidwrighton commented:
@MihaZupan @sbomer @jkoritzinsky all of you had good feedback. I believe I've addressed it all, and things are working again. Could you please take another look at this?

sbomer reviewed (approved):
A couple more nits, otherwise LGTM. Thanks!

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into IntrinsicsAnalyzerAndFixes

davidwrighton committed:
With the analyzer, negative checks are problematic, so re-order some of them. Also we needed some new pragma warning disables to make the new logic work

davidwrighton committed:
Code review feedback

MihaZupan reviewed (commented):
Just questions around where the annotations are needed

jkoritzinsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilerContext.cs:
```diff
@@ -67,6 +69,8 @@ public ReadyToRunCompilerContext(TargetDetails details, SharedGenericsMode gener
             }
         }
 
+        public InstructionSetSupport InstructionSetSupport { get { return _instructionSetSupport; } }
```
nit: Auto-implemented get-only property instead of manual read-only field + property?

jkoritzinsky commented on src/libraries/System.Private.CoreLib/tests/IntrinsicsInSystemPrivatecoreLibAnalyzer.Tests/IntrinsicsInSystemPrivateCoreLib.Tests.csproj:
```diff
@@ -0,0 +1,24 @@
+
+  <PropertyGroup>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
+    <Platforms>AnyCPU;x64</Platforms>
```
I think we can remove these properties.

```suggestion
```

jkoritzinsky commented on src/libraries/System.Private.CoreLib/gen/IntrinsicsInSystemPrivateCoreLibAnalyzer.cs:
```diff
@@ -0,0 +1,693 @@
+                return null;
+        }
+
+        private static INamedTypeSymbol[] GatherAndConditions(SemanticModel model, ExpressionSyntax expressionToDecompose)
```
This would be better done operating on `IOperation` instances than syntax, but that can be a later refactoring (as this does work).

davidwrighton commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs:
```diff
@@ -2444,6 +2444,10 @@ public static Vector128<sbyte> Shuffle(Vector128<sbyte> vector, Vector128<sbyte>
         /// On hardware with <see cref="AdvSimd.Arm64"/> or <see cref="PackedSimd"/> support, this method behaves the same as Shuffle.</remarks>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Ssse3))]
+        [CompExactlyDependsOn(typeof(AdvSimd))]
```
Sortof. `ShuffleUnsafe` is pretty much impossible to attribute correctly. As an api which behaves differently based on the particular architecture it runs on, the safest thing to do is just cover it in various CompExactlyDependsOn attributes, and then deal with the downstream effects.

jkoritzinsky commented on src/libraries/System.Private.CoreLib/gen/IntrinsicsInSystemPrivateCoreLibAnalyzer.cs:
```diff
@@ -0,0 +1,693 @@
+        }
+
+        private static INamespaceSymbol? SymbolToNamespaceSymbol(ISymbol symbol)
+        {
+            if (symbol.ContainingType != null)
+                return SymbolToNamespaceSymbol(symbol.ContainingType);
+
+            if (symbol.ContainingNamespace != null)
+            {
+                return symbol.ContainingNamespace;
+            }
+
+            return null;
+        }
```
Is there a reason we can't just use `ContainingNamespace` here? It should work for all symbols, not just type symbols IIRC.

```suggestion
        {
            return symbol.ContainingNamespace;
        }
```

davidwrighton commented on src/libraries/System.Private.CoreLib/src/System/SearchValues/ProbabilisticMap.cs:
```diff
@@ -165,6 +168,11 @@ private static Vector128<byte> ContainsMask16Chars(Vector128<byte> charMapLower,
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Sse2))]
+        [CompExactlyDependsOn(typeof(Ssse3))]
+        [CompExactlyDependsOn(typeof(AdvSimd))]
```
It needs Sse2. The attribution is not whether or not Ssse3 implies Sse2 or not, but rather that the behavior of Sse2.IsSupported will change how the function executes even if the meaning of Ssse3.IsSupported does not change in meaning.

davidwrighton replied:
The PackedSimd check is to deal with use of ShuffleUnsafe.

davidwrighton commented on src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf16Utility.Validation.cs:
```diff
@@ -130,7 +130,7 @@ internal static unsafe partial class Utf16Utility
                         // and 3-byte) elements.
 
-                        if (AdvSimd.IsSupported)
+                        if (AdvSimd.Arm64.IsSupported)
```
This function behaves the same way regardless of what itnrinsics are available for use at runtime. In particular, it has a fallback path for nonintrinsic execution, and all use of intrinsics is protected by positive `IsSupported` checks.

davidwrighton commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs:
```diff
@@ -35,37 +35,46 @@ public static unsafe bool CanUsePackedIndexOf<T>(T value)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Sse2))]
```
Yes. All of the helpers which rely on potentially variant behavior without a complete set of IsSupported checks need annotation.

davidwrighton commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Packed.cs:
```diff
@@ -447,7 +482,16 @@ private static int IndexOfAny<TNegator>(ref short searchSpace, short value0, sho
                     Vector128<byte> packedValue0 = Vector128.Create((byte)value0);
                     Vector128<byte> packedValue1 = Vector128.Create((byte)value1);
 
+#pragma warning disable IntrinsicsInSystemPrivateCoreLibConditionParsing // A negated IsSupported condition isn't parseable by the intrinsics analyzer, but in this case, it is only used in combination
```
Feel free to refactor after I check in. This comment style has made it much easier to get the correct pragmas into place as I look at lots of code, but that need will evaporate once everything is in.

davidwrighton commented on src/libraries/System.Private.CoreLib/src/System/SearchValues/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -827,6 +866,9 @@ internal static int LastIndexOfAnyVectorizedAnyByte<TNegator>(ref byte searchSpa
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Ssse3))]
+        [CompExactlyDependsOn(typeof(AdvSimd))]
+        [CompExactlyDependsOn(typeof(PackedSimd))]
```
I assume so. I didn't just add stuff unless my tooling detected it was necessary.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/SearchValues/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -815,6 +851,9 @@ internal static int LastIndexOfAnyVectorizedAnyByte<TNegator>(ref byte searchSpa
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(Sse2))]
+        [CompExactlyDependsOn(typeof(AdvSimd))]
+        [CompExactlyDependsOn(typeof(PackedSimd))]
```
> Effectively, we need to a be aware of all possible inlining behaviors.

Thanks, this part wasn't obvious to me (and pretty much answers all the other questions as well).

davidwrighton committed:
Update src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

davidwrighton committed:
Update src/libraries/System.Private.CoreLib/tests/IntrinsicsInSystemPrivatecoreLibAnalyzer.Tests/IntrinsicsInSystemPrivateCoreLib.Tests.csproj

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

davidwrighton committed:
Update src/libraries/System.Private.CoreLib/gen/IntrinsicsInSystemPrivateCoreLibAnalyzer.cs

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

davidwrighton committed:
Update IntrinsicsInSystemPrivateCoreLibAnalyzer.cs

davidwrighton committed:
Use auto-implemented get-only property

davidwrighton commented on src/libraries/System.Private.CoreLib/tests/IntrinsicsInSystemPrivatecoreLibAnalyzer.Tests/CSharpAnalyzerVerifier`1+Test.cs:
```diff
@@ -0,0 +1,24 @@
+        {
+            public Test()
+            {
+                SolutionTransforms.Add((solution, projectId) =>
+                {
+                    var compilationOptions = solution.GetProject(projectId).CompilationOptions;
+                    solution = solution.WithProjectCompilationOptions(projectId, compilationOptions);
+
+                    return solution;
+                });
```
Honestly, I have no idea. It came from the default make an analyzer template in VS, and doesn't seem to hurt anything.

jkoritzinsky replied:
Weird. I'm surprised that this is in the template. It's not important one way or the other.

davidwrighton commented on src/libraries/System.Private.CoreLib/gen/IntrinsicsInSystemPrivateCoreLibAnalyzer.cs:
```diff
@@ -0,0 +1,693 @@
+            }
+
+            bool methodNeedsProtectionWithIsSupported = false;
+#pragma warning disable RS1024
```
Done
